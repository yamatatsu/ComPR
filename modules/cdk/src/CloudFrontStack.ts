import * as cdk from "@aws-cdk/core"
import * as cloudfront from "@aws-cdk/aws-cloudfront"
import * as s3 from "@aws-cdk/aws-s3"

type Props = cdk.StackProps & {}

export class CloudFrontStack extends cdk.Stack {
  constructor(parent: cdk.Construct, id: string, props: Props) {
    super(parent, id, props)

    const bucket = new s3.Bucket(this, `bucket`, {
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    })
    const oai = new cloudfront.OriginAccessIdentity(this, `oai`)

    new cloudfront.CloudFrontWebDistribution(
      this,
      "CloudFrontWebDistribution",
      {
        viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
        httpVersion: cloudfront.HttpVersion.HTTP2,
        originConfigs: [
          {
            s3OriginSource: {
              s3BucketSource: bucket,
              originAccessIdentity: oai,
            },
            behaviors: [
              {
                pathPattern: "index.html",
                minTtl: cdk.Duration.seconds(0),
                maxTtl: cdk.Duration.seconds(0),
                defaultTtl: cdk.Duration.seconds(0),
              },
              {
                isDefaultBehavior: true,
                minTtl: cdk.Duration.seconds(0),
                maxTtl: cdk.Duration.days(365),
                defaultTtl: cdk.Duration.days(1),
              },
            ],
          },
        ],
        errorConfigurations: [
          {
            errorCachingMinTtl: 0,
            errorCode: 403,
            responseCode: 200,
            responsePagePath: "/index.html",
          },
          {
            errorCachingMinTtl: 0,
            errorCode: 404,
            responseCode: 200,
            responsePagePath: "/index.html",
          },
        ],
      },
    )

    new cdk.CfnOutput(this, "bucket-name", {
      value: bucket.bucketName,
    })
  }
}
