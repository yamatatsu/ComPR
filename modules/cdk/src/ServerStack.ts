import * as cdk from "@aws-cdk/core"
import * as lambda from "@aws-cdk/aws-lambda"
import * as apigateway from "@aws-cdk/aws-apigatewayv2"

type Props = cdk.StackProps & {
  client_id: string
  client_secret: string
  redirect_uri: string
  code: lambda.Code
  layerCode: lambda.Code
  allowOrigins: string
}

export class ServerStack extends cdk.Stack {
  constructor(parent: cdk.Construct, id: string, props: Props) {
    super(parent, id, props)

    const {
      client_id,
      client_secret,
      redirect_uri,
      code,
      layerCode,
      allowOrigins,
    } = props

    const nodeModulesLayer = new lambda.LayerVersion(this, "NodeModulesLayer", {
      code: layerCode,
      compatibleRuntimes: [lambda.Runtime.NODEJS_12_X],
    })
    const authTokenHandler = new lambda.Function(this, "authTokenHandler", {
      code,
      layers: [nodeModulesLayer],
      handler: "authToken.handler",
      runtime: lambda.Runtime.NODEJS_12_X,
      environment: { client_id, client_secret, redirect_uri },
      tracing: lambda.Tracing.ACTIVE,
    })

    const httpApi = new apigateway.HttpApi(this, "httpApi", {
      apiName: `cooprate-httpApi-${id}`,
      corsPreflight: {
        allowCredentials: false,
        allowHeaders: ["content-type"],
        allowMethods: [apigateway.HttpMethod.GET, apigateway.HttpMethod.POST],
        allowOrigins: [allowOrigins],
      },
    })

    httpApi.addRoutes({
      path: "/auth/token",
      methods: [apigateway.HttpMethod.POST],
      integration: new apigateway.LambdaProxyIntegration({
        handler: authTokenHandler,
      }),
    })

    new cdk.CfnOutput(this, "api-url", {
      value: httpApi.url ?? "no url",
    })
  }
}
