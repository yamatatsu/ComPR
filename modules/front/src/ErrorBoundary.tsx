import React from "react"

type Props = {
  children: React.ReactElement
}
type State = {
  hasError: boolean
}
export class ErrorBoundary extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = { hasError: false }
  }

  static getDerivedStateFromError() {
    return { hasError: true }
  }

  componentDidCatch(error: any, errorInfo: any) {
    // TODO: use error tracing service
    console.group("`ErrorBoundary` has caught a error.")
    console.error(error)
    console.error(errorInfo)
    console.groupEnd()
  }

  render() {
    if (this.state.hasError) {
      return (
        <div>
          <h3>エラーが発生しました。再描画をお試しください。</h3>
          <button onClick={() => location.reload()}>再描画する</button>
        </div>
      )
    }

    return this.props.children
  }
}
