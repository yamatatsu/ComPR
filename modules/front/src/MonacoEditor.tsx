import React, { useRef, useEffect } from "react"
import * as monaco from "monaco-editor"
import { MonacoMarkdownExtension } from "monaco-markdown"

type Props = { code: string }
export default function MonacoEditor(props: Props) {
  const { code } = props
  const ref = useRef<HTMLDivElement>(null)

  useEffect(() => {
    if (ref.current) {
      const model = monaco.editor.createModel(code, "markdown")
      model.updateOptions({ tabSize: 2 })
      const ed = monaco.editor.create(ref.current, {
        model,
        language: "markdown",
        fontSize: 14,
        roundedSelection: false,
        scrollBeyondLastLine: false,
        theme: "vs",
        minimap: { enabled: false },
      })
      // @ts-ignore
      new MonacoMarkdownExtension().activate(ed)
    }
  }, [ref])
  return <div style={{ width: "100%", height: "100%" }} ref={ref}></div>
}
