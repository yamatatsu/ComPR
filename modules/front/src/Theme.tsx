import React, { FunctionComponent } from "react"
import { BaseStyles, theme } from "@primer/components"
import { ThemeProvider } from "styled-components"

export const Theme: FunctionComponent = ({ children }) => {
  return (
    <BaseStyles>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </BaseStyles>
  )
}
