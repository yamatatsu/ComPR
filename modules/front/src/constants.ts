const redirect_uri = process.env.redirect_uri
const client_id = process.env.client_id
const api_origin = process.env.api_origin

export const getConstants = () => {
  if (!redirect_uri) throw new Error("No redirect_uri is set in .env")
  if (!client_id) throw new Error("No client_id is set in .env")
  if (!api_origin) throw new Error("No api_origin is set in .env")

  return { client_id, api_origin, redirect_uri }
}
