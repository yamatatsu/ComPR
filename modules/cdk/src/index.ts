import * as cdk from "@aws-cdk/core"
import * as lambda from "@aws-cdk/aws-lambda"
import { CloudFrontStack } from "./CloudFrontStack"
import { ServerStack } from "./ServerStack"

const {
  client_id,
  client_secret,
  redirect_uri,
  dev_client_id,
  dev_client_secret,
  dev_redirect_uri,
} = process.env
if (!client_id) throw new Error("no client_id in `process.env`")
if (!client_secret) throw new Error("no client_secret in `process.env`")
if (!redirect_uri) throw new Error("no redirect_uri in `process.env`")
if (!dev_client_id) throw new Error("no dev_client_id in `process.env`")
if (!dev_client_secret) throw new Error("no dev_client_secret in `process.env`")
if (!dev_redirect_uri) throw new Error("no dev_redirect_uri in `process.env`")

const env = { region: "ap-northeast-1" }

const app = new cdk.App()

const code = new lambda.AssetCode(`${__dirname}/../../lambda/dist`)
const layerCode = new lambda.AssetCode(`${__dirname}/../layer`)

new ServerStack(app, "ServerStack", {
  client_id,
  client_secret,
  redirect_uri,
  code,
  layerCode,
  allowOrigins: "https://d2gteuwc4ok948.cloudfront.net",
  env,
})

new ServerStack(app, "DevServerStack", {
  client_id: dev_client_id,
  client_secret: dev_client_secret,
  redirect_uri: dev_redirect_uri,
  code,
  layerCode,
  allowOrigins: "http://localhost:8080",
  env,
})

new CloudFrontStack(app, "CloudFrontStack", { env })
