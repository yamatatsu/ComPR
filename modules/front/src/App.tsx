import React, { useState } from "react"
import {
  BrowserRouter,
  Route,
  Redirect,
  Switch,
  useParams,
} from "react-router-dom"
import ApolloClient, {
  InMemoryCache,
  IntrospectionFragmentMatcher,
} from "apollo-boost"
import { ApolloProvider } from "@apollo/react-hooks"
import { ErrorBoundary } from "./ErrorBoundary"
import { Theme } from "./Theme"

import introspectionQueryResultData from "../schema/fragmentTypes.json"
import { LandingPage } from "./components/LandingPage"
import { GithubLogin, GithubLoginAndRedirect } from "./components/GithubLogin"
import { GithubCallback } from "./components/GithubCallback"
import { Redirecter } from "./components/Redirecter"
import { RepoList } from "./components/RepoList"
import { Explorer } from "./components/Explorer"
import { NewEditor, EditEditor } from "./components/Editor"
import { NotFound } from "./components/NotFound"

export function App() {
  const [token, setToken] = useState<string | null>(
    process.env.token_for_debug ?? null,
  )

  if (!token) {
    return (
      <ErrorBoundary>
        <Theme>
          <BrowserRouter>
            <Switch>
              <Route exact path="/" component={LandingPage} />
              <Route exact path="/login" component={GithubLogin} />
              <Route exact path="/login/callback">
                <GithubCallback setToken={setToken} />
              </Route>
              <Route path="*" component={GithubLoginAndRedirect} />
            </Switch>
          </BrowserRouter>
        </Theme>
      </ErrorBoundary>
    )
  }

  const client = new ApolloClient({
    cache: new InMemoryCache({
      fragmentMatcher: new IntrospectionFragmentMatcher({
        introspectionQueryResultData,
      }),
    }),
    uri: "https://api.github.com/graphql",
    headers: {
      authorization: `Bearer ${token}`,
    },
  })

  return (
    <ErrorBoundary>
      <ApolloProvider client={client}>
        <Theme>
          <BrowserRouter>
            <Switch>
              <Route exact path="/" component={RepoList} />
              <Route exact path="/login/callback" component={Redirecter} />
              <Route
                path="/:owner/:repo/:branch/new"
                component={() => {
                  const { owner, repo, branch } = useParams()
                  return <NewEditor owner={owner} repo={repo} branch={branch} />
                }}
              />
              <Route
                path="/:owner/:repo/:branch/edit"
                component={() => {
                  const { owner, repo, branch } = useParams()
                  return (
                    <EditEditor owner={owner} repo={repo} branch={branch} />
                  )
                }}
              />
              <Route
                path="/:owner/:repo/:branch/list"
                component={() => {
                  const { owner, repo, branch } = useParams()
                  return <Explorer owner={owner} repo={repo} branch={branch} />
                }}
              />
              <Route exact path="/not-found" component={NotFound} />
              <Redirect from="*" to="/not-found" />
            </Switch>
          </BrowserRouter>
        </Theme>
      </ApolloProvider>
    </ErrorBoundary>
  )
}
