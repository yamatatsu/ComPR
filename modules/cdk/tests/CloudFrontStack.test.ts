import { SynthUtils } from "@aws-cdk/assert"
import { Stack } from "@aws-cdk/core"
import { CloudFrontStack } from "../src/CloudFrontStack"

test("CloudFrontStack snapshot test", () => {
  const stack = new Stack()
  const env = { region: "ap-northeast-1", account: "xxxxxs" }

  const target = new CloudFrontStack(stack, "testCloudFrontStack", {
    env,
  })

  expect(SynthUtils.toCloudFormation(target)).toMatchSnapshot()
})
