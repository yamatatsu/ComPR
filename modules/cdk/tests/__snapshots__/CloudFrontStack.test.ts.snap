// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudFrontStack snapshot test 1`] = `
Object {
  "Outputs": Object {
    "bucketname": Object {
      "Value": Object {
        "Ref": "bucket43879C71",
      },
    },
  },
  "Resources": Object {
    "CloudFrontWebDistributionCFDistribution9F34B88B": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "CacheBehaviors": Array [
            Object {
              "AllowedMethods": Array [
                "GET",
                "HEAD",
              ],
              "CachedMethods": Array [
                "GET",
                "HEAD",
              ],
              "Compress": true,
              "DefaultTTL": 0,
              "ForwardedValues": Object {
                "Cookies": Object {
                  "Forward": "none",
                },
                "QueryString": false,
              },
              "MaxTTL": 0,
              "MinTTL": 0,
              "PathPattern": "index.html",
              "TargetOriginId": "origin1",
              "ViewerProtocolPolicy": "redirect-to-https",
            },
          ],
          "CustomErrorResponses": Array [
            Object {
              "ErrorCachingMinTTL": 0,
              "ErrorCode": 403,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
            Object {
              "ErrorCachingMinTTL": 0,
              "ErrorCode": 404,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
          ],
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "DefaultTTL": 86400,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "MaxTTL": 31536000,
            "MinTTL": 0,
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "bucket43879C71",
                  "RegionalDomainName",
                ],
              },
              "Id": "origin1",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "oai7A77E975",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "CloudFrontDefaultCertificate": true,
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "bucket43879C71": Object {
      "DeletionPolicy": "Delete",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "bucketPolicy638F945D": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "bucket43879C71",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "oai7A77E975",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "bucket43879C71",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "bucket43879C71",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "oai7A77E975": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
  },
}
`;
